lib/pltr/v2/reducers/root.js

```diff
@@ -39,6 +43,63 @@ const root = (dataRepairers) => (state, action) => {
         nextCardId: nextId(state.cards),
       })
 
+    case ADD_LINES_FROM_TEMPLATE: {
+      // cards from the template need to know the new ids of lines (and sometimes beats) from the template
+      // FOR LINES:
+      // the strategy here is to use the state's next id value + the template id's current value
+      // the card reducer will have access to the state's next id value so it will be able to determine the correct id
+      // FOR BEATS:
+      // cards will use the cardToBeatIdMap to use the current book's beat ids
+      // but if more beats are needed, they will be created with subsequent ids
+      // ALSO:
+      // TODO:
+      // ignore beats unless there are > 1 hierarchyLevels
+      // add cards to leaf nodes (beats) unless there are > 1 hierarchyLevels
+      const bookId = state.ui.currentTimeline
+      let nextIdForBeats = nextBeatId(state.beats)
+      let beatTree = cloneDeep(state.beats[bookId])
+      let createdNewBeats = false
+      // make a card -> beatId mapping (beatId is from existing beats … augmented with new ones)
+      const beatPositionMap = Object.keys(beatTree.index).reduce((acc, id) => {
```

Comment:
I think we should use `tree.reduce` instead so that the root reducer is insulated from the implementation of beats.

Reply:
On second thought: we have a helper function that flattens a tree in sorted order already:
`helpers/beats.js:beatsByPosition`.
This function insulates the reducer from another detail: the fact that beats have a `position` and simplifies this function: it would become something like:
```
beatsByPosition(() => true)(beatTree).map(({ id }) => id)
```

Reply:
It doesn't make much sense for deeper hierarchies because you'd overlay the cards over the hierarchy regardless of level.